
## World Happiness Report 2023

### Branson Enani

Througout this project we will be working with data collected for the World Happiness Report in 2023. This data set contains information regarding happiness and wellbeing spanning all countries from the year 2008 to the present. 



Before we begin to tackle the questions we will be facing, we should first get familiar with the data. The purpose of this data set is to determine the general happines of individuals around the world and to determine what causes the biggest impact on happiness in an individual. This data comes from a few sources: the Gallup World Poll (GWP), the World Development Indicators (WDI), the Penn World Table 10.01, and the World Health Organization's Global Health Observatory data repository. These sources are all administrative in nature, so it is safe to say that this data set is administrative data - that is, data collected by organizations for their operations which is continuously updated. This data was collected from individuals all around the world, so the population is the global population of humans. However, it is important to recognize the bias this data may present. Who had access to the poll? For poorer countries or countries with very high populations, the poll may have only have been accessed by people with access to technology or otherwise. Data could have been collected by only people who were rich enough to partake as well. We need to keep this in mind when analyzing the data. Each observation is a country's average values for each variable based upon each individual response. The variables are as follows:

### Country Name and year

These variable names are very strightforward. After briefly inspecting the data, we found that there are 165 different countries and the range of years is from 2005 to 2022.

### Life Ladder

This is the overall subjective happiness score. It is the national average response to the question of life evaluations. The reason it is 
called Life ladder is because it is a response the the question “Please imagine a ladder, with steps numbered from 0 at the
bottom to 10 at the top. The top of the ladder represents the best possible life for you and the bottom of the ladder represents the worst possible life for you.
On which step of the ladder would you say you personally feel you stand at this time?”

### Log Gdp Per Capita

The log Gdp Per capita applies the log function to the economic output of the country per person. It essentially describes the economic strength of the country and divides this among the population to estimate the individual economic well-being of citizens. As stated in the statistical appendix, "The statistics of GDP per capita (variable name gdp) in purchasing power parity (PPP) at constant 2017 international dollar prices are from World Development Indicators (WDI, version 17, metadata last updated on Jan 22, 2023). The GDP figures for Taiwan, Syria, Palestine, Venezuela, Djibouti and Yemen are from
the Penn World Table 10.01."

### Social Support

The social support variable is calculated as a proportion of people answering yes to the question of whether they feel they have social support or not. 
Specifically, this response applies to the question of "If you were in trouble, do you have relatives or friends you can count on to help you
whenever you need them, or not?”


### Healthy Life expectancy at birth

Healthy Life Expectancy at birth is fairly self-explanatory. Healthy Life expectancy, as defined by the World Health Organization (WHO) as "Average number of years that a person can expect to live in "full health" by taking into account years lived in less than full health due to disease and/or injury." The data in this table was taken from the WHO from the years of 2000, 2010, 2015, and 2019. The remaining years of healthy life expectancy are calculated via extrapolation and interpolation 

### Freedom to make life choices

This variable of Freedom to make life choices is calcuated in a similar manner to the social support variable; it is a proportion of "yes" responses. The question posed for this variable is “Are you satisfied or dissatisfied with your freedom to choose what you do with your life?”

### Generosity

Generosity in the context of this dataset is a measurement of donating to charity. Specifically it is the proportion of people answering yes to the question "have you donated to charity in the last month?" 

### Percpetions of Corruption

This variable is the average of reponses to two different questions instead of just 1 like some of the previous questions. A no corresponds to 0 and a yes corresponds to 1. The two questions are:
1)  “Is corruption widespread throughout the government or not?”
2) “Is corruption widespread within businesses or not?”

### Positive Affect

Positive Affect is basically a variable describing the positive moods or emotional states of people on average. There are three specific affect measures 
that comprise this variable: laugh, enjoyment, and doing interesting things. By answering questions about experiences of each of these affect measures in the previous day, an average can be calcuated from averaging the yes and no responses.

### Negative Affect

Negative Affect measures more negative or unpleasant emotional states. The three negative affect measures are: worry, sadnesss and anger. By answering questions about experiences of each of these affect measures in the previous day, an average can be calcuated from averaging the yes and no responses.

The following is a table which gives an organized visual representation of these variables

Name | Variable description | Type | Units of measurement
---|---|---|---
Country name | The name of the country | Nominal | None 
year | Year of collection | Numeric | None
Life Ladder | Average happiness rating from 1 - 10 with 10 being the highest level of happiness | Numeric | None
Log GDP per capita | The gross domestic product of a country on a log scale | Numeric | PPD
Social support | Average proportion of responses to whether or not someone has a social group to go to in times of need, with 0 meaning no and 1 meaning yes | Numeric | None
Healthy life expectancy at birth | Average life expectancy of a healthy individual at the time of birth | Numeric | Year
Freedom to make life choices | Average proportion of repsonses to whether or not someone feels they have freedom over their choices, with 0 meaning no and 1 meaning yes | Numeric | None
Generosity | Average proportion of responses to whether or not someone has donated to charity in the past month, with 0 meaning no and 1 meaning yes | Numeric | None
Perceptions of corruption | Average proportion of responses to whether or not someone feels their country is corrupt, with 0 meaning no and 1 meaning yes | Numeric | None
Positive affect | Average proportion of repsonses to 3 different positive affect experiences in the last day, with 0 meaning no and 1 meaning yes | Numeric | None
Negative affect | Average proportion of repsonses to 3 different negative affect experiences in the last day, with 0 meaning no and 1 meaning yes | Numeric | None

All in all, this data set contains the mean value of a variety of different poll questions answered by people around the world about their country and their own emotions and actions. Throughout this project, I will be trying to find the link between these answers and the Life Ladder - also known as the happiness rating. I want to find which variables influence Life Ladder the most, which influence Life Ladder the least, and the way these variables differ by country, year, and GDP.  

# Part 2 : Questions of Interest

Throughout this project I will be posing a few questions in an attempt to understand more about this data set and the levels of happiness around the world. The questions I am asking are as follows:

* (i) What is the relationship between Life Ladder and Log GDP per capita? Does a high Log GDP per capita cause a higher Life Ladder or lower Life Ladder?
* (ii) Are there any countries who have gotten at least one point happier from the years 2008 to 2022? Are there any that have gotten at least one point sadder?
* (iii) Does Life Ladder seem to differ between different continents? What continent has the lowest and highest Life Ladder?
* (iv) Does Life Expectancy influence Life Ladder? Which country has the lowest Life Expectancy and which has the highest? Compare their Life Ladder values. Which continents?
* (v) How do social support, freedom to make life choices, generosity, perceptions of corruption, positive affect, and negative affect correlate with Life Ladder? Do any of them correlate with each other? Why do you think this is?

These questions all fall underneath the main umbrella of my analysis, the question:

## Are we able to predict the Life Ladder of a certain country based on given values of the various predictors? What has the most influence on Life Ladder?

We will be exploring the answer to these questions in the sections below. To sufficiently answer these questions, we will have to build a variety of plots and graphics, as well as perform a regression analysis. This will tell us whether or not the data is linear and from there we will be able to answer the question of whether or not Life Ladder is predictable based on certain predictor values.

# Part 3: Data Analysis
## (i) Exploratory Data Analysis

The first step in any analysis is to import the necessary packages we will be using for our analysis, and of course, the data itself.

In [62]:
# install pycountry_convert
!pip install pycountry_convert

In [63]:
# import packages
import pandas as pd
import numpy as np
import altair as alt
from scipy import linalg
import statsmodels.api as sm
import pycountry_convert as pc
# disable row limit for plotting
alt.data_transformers.disable_max_rows()
# uncomment to ensure graphics display with pdf export
# alt.renderers.enable('mimetype')


# read in data
happiness = pd.read_csv('~/pstat100-content/Course_Project/whr-2023.csv')

happiness.head()

,Country name,year,Life Ladder,Log GDP per capita,Social support,Healthy life expectancy at birth,Freedom to make life choices,Generosity,Perceptions of corruption,Positive affect,Negative affect
0,Afghanistan,2008,3.724,7.350,0.451,50.5,0.718,0.168,0.882,0.414,0.258
1,Afghanistan,2009,4.402,7.509,0.552,50.8,0.679,0.191,0.850,0.481,0.237
2,Afghanistan,2010,4.758,7.614,0.539,51.1,0.600,0.121,0.707,0.517,0.275
3,Afghanistan,2011,3.832,7.581,0.521,51.4,0.496,0.164,0.731,0.480,0.267
4,Afghanistan,2012,3.783,7.661,0.521,51.7,0.531,0.238,0.776,0.614,0.268


Let's first work on answering the first exploratory analysis question:

What is the relationship between Life Ladder and Log GDP per capita? Does a high Log GDP per capita cause a higher Life Ladder or lower Life Ladder?

To do this, we can simply create a scatterplot and observe the general trend between Life Ladder and Log GDP per capita.

In [64]:
##### EXPLORATORY ANALYSIS #####
######### QUESTION ONE #########
################################

# (i) What is the relationship between Life Ladder and Log GDP per capita?
# Does a high Log GDP per capita cause a higher Life Ladder or lower Life Ladder? Group by year.

# Create scatterplot of Life Ladder vs Log GDP per capita
ll_vs_gdp = alt.Chart(happiness, title = 'Life Ladder vs Log GDP per capita').mark_point(opacity = 0.8).encode(
    y = alt.Y('Life Ladder', scale = alt.Scale(domain = [2, 8.2])),
    x = alt.X('Log GDP per capita', scale = alt.Scale(domain = [5, 12])),
    color = 'year'
)

ll_vs_gdp

alt.Chart(...)

There is a clear positively correlated relationship between Life Ladder and Log GDP per capita. Furthermore, it looks like as the years get more recent, life ladder is increasing, as the clumps of dark blue points seem to be conglomerated around the top right of the graph. 

There also seems to be a few outliers - specifically, the three points with very low log GDP per capita but decently high Life Ladder in comparison. Let's find these outliers.

In [65]:
# Find the low Log GDPS per capita
happiness.sort_values('Log GDP per capita', ascending = True).head(3)

,Country name,year,Life Ladder,Log GDP per capita,Social support,Healthy life expectancy at birth,Freedom to make life choices,Generosity,Perceptions of corruption,Positive affect,Negative affect
2134,Venezuela,2019,5.081,5.527,0.888,64.400,0.626,0.128,0.839,0.730,0.351
2133,Venezuela,2018,5.006,5.935,0.887,64.575,0.611,0.072,0.828,0.723,0.374
2132,Venezuela,2017,5.071,5.943,0.896,64.750,0.636,0.054,0.844,0.697,0.363


The three outliers are shared by the same country in three consecutive years - Venezuela in 2017, 2018, and 2019. The economy in Venezuela has been rough in the past years, so this could possibly be an explanation. As the economy struggles, people still find ways to be happy. Social support and positive affect are both relatively high, which definitely influences Life Ladder in some way.

The next question we will be answering is:

Are there any countries who have gotten at least one point happier from the years 2008 to 2022? Are there any that have gotten at least one point sadder?

To do this we must first filter the data so we only have the Life Ladder for each country in the years 2008 and 2022. Then, we must calculate the difference between the two and filter the dataset so only countries with a difference greater than 1 or less than -1 are shown.

In [66]:
##### EXPLORATORY ANALYSIS #####
######### QUESTION TWO #########
################################

# (ii) Are there any countries who have drastically different Life Ladders between the years 2008 - 2022?
# If yes, what may have caused this? If not, why has it remained stagnant?

# Filter data for years 2008 and 2022
filtered_data = happiness[(happiness['year'] == 2008) | (happiness['year'] == 2022)]

# Group the data by country and calculate the difference
difference_data = filtered_data.pivot(index = 'Country name', columns = 'year', values = 'Life Ladder')
difference_data['difference'] = difference_data[2022] - difference_data[2008]
difference_data.reset_index(inplace = True)

# Find countries that have either increased or decreased by a point or more between the years 2008 and 2022
big_diff_data = difference_data[(difference_data['difference'] > 1) | (difference_data['difference'] < -1)]

# Rename index
big_diff_data.columns.name = None
big_diff_data.reset_index(inplace = True, drop = True)

big_diff_data

,Country name,2008,2022,difference
0,Afghanistan,3.724,1.281,-2.443
1,Bangladesh,5.052,3.408,-1.644
2,Botswana,5.451,3.435,-2.016
3,Congo (Brazzaville),3.820,5.805,1.985
4,El Salvador,5.191,6.492,1.301
5,Georgia,4.156,5.293,1.137
6,India,5.146,3.930,-1.216
7,Lebanon,4.595,2.352,-2.243
8,Lithuania,5.554,7.038,1.484
9,Mongolia,4.493,5.788,1.295


Now that we have the information for what countries had at least one point of change, we can graph the Life Ladder trend lines per year for countries with a negative change and for countries with a positive change.

In [67]:
# Filter data for negative countries and positive countries
plot_negative = happiness[(happiness['Country name'] == 'Afghanistan') | (happiness['Country name'] == 'Bangladesh')
                      | (happiness['Country name'] == 'Botswana')
                      | (happiness['Country name'] == 'India') | (happiness['Country name'] == 'Lebanon')]

plot_positive = happiness[(happiness['Country name'] == 'Lithuania') | (happiness['Country name'] == 'Mongolia')
                      | (happiness['Country name'] == 'Nepal') | (happiness['Country name'] == 'Nicaragua')
                      | (happiness['Country name'] == 'Taiwan Province of China') | (happiness['Country name'] == 'Togo')
                      | (happiness['Country name'] == 'Uruguay') | (happiness['Country name'] == 'Congo (Brazzaville)')
                      | (happiness['Country name'] == 'El Salvador') | (happiness['Country name'] == 'Georgia')]

# Filter data for years 2008 and onwards
plot_positive = plot_positive[plot_positive['year'] >= 2008]
plot_negative = plot_negative[plot_negative['year'] >= 2008]

# Create a plot for negative change 
negative = alt.Chart(plot_negative, title = 'Life Ladder per Year for Countries with Life Ladder Decrease').mark_line().encode(
    y = alt.Y('Life Ladder', scale = alt.Scale(domain = [1, 6])),
    x = alt.X('year', scale = alt.Scale(domain = [2008, 2022])),
    color = 'Country name'
)

negative

alt.Chart(...)

The countries that experienced a negative change are Afghanistan, Bangladesh, Botswana, India, and Lebanon. The order of countries from biggest drop to smallest drop in Life Ladder is Afghanistan, Lebanon, Botswana, Bangladesh, and India. 

Based on the countries listed, I can make a few guesses as to why Life Ladder has decreased. Afghanistan has been trending downwards consistently due to multiple wars in the country. Lebanon was doing well, but in 2018 dropped dramatically until 2021. This is most likely due to the outcome of the 2018 Lebanese election and the consequences that followed. Botswana has been constantly in a state of civil unrest, especially since the recent war starting in 2021. Bangladesh has definitely gotten worse because of the Coronavirus and the huge impact it had on the country - this is most likely the reason for the huge dip in 2020 - 2022. I would guess the same for India, but India has actually increased a bit in recent years. India's change is most likely due to the country's already poor stature, and possibly even Climate Change and overpopulation.

Let's take a look at the countries with a positive change!

In [68]:
# Create a plot for positive change 
positive = alt.Chart(plot_positive, title = 'Life Ladder per Year for Countries with Life Ladder Increase').mark_line().encode(
    y = alt.Y('Life Ladder', scale = alt.Scale(domain = [2.5, 7.5])),
    x = alt.X('year', scale = alt.Scale(domain = [2008, 2022])),
    color = 'Country name'
)

positive

alt.Chart(...)

The countries that changed positively by one point or more as as follows, in order from highest change to smallest change: 

Congo, El Salvador, Georgia, Lithuania, Mongolia, Nepal, Nicaragua Taiwan, Togo, Uruguay.



In [69]:
# It would also be interesting to see how their GDPs changed per year. Let's find out
# Create a plot for negative change GDP
negative_gdp = alt.Chart(plot_negative, title = 'Log GDP per capita per Year for Countries with Life Ladder Decrease').mark_line().encode(
    y = alt.Y('Log GDP per capita', scale = alt.Scale(domain = [7, 10])),
    x = alt.X('year', scale = alt.Scale(domain = [2008, 2022])),
    color = 'Country name'
)

negative_gdp

alt.Chart(...)

In [70]:
# Create a plot for positive change GDP
positive_gdp = alt.Chart(plot_positive, title = 'Log GDP per capita per Year for Countries with Life Ladder Increase').mark_line().encode(
    y = alt.Y('Log GDP per capita', scale = alt.Scale(domain = [7, 11])),
    x = alt.X('year', scale = alt.Scale(domain = [2008, 2022])),
    color = 'Country name'
)

positive_gdp

alt.Chart(...)

In [71]:
# Positive countries : Congo, El Salvador, Georgia, Lithuania, Mongolia, Nepal, Nicaragua
# Taiwan, Togo, Uruguay

# Negative countries : Afghanistan, Bangladesh, Botswana, India, Lebanon

# For GDPS, for nearly every positive country except Congo there was an increase.
# However, the same cant be said for negative countries. Afhganistan and Lebanon decreased slightly, but the rest 
# actually increased.

# Huge negative dip for lebanon, bangladesh, and afghanistan. Wars for Leb and Afghanistan, Bangladesh environmental ? look it up
# Highest positive jump was Congo. Why? No clue!

In [72]:
##### EXPLORATORY ANALYSIS #####
######## QUESTION THREE ########
################################

# (iii) Does Life Ladder seem to differ between different continents? What continent has the lowest and highest Life Ladder?

# Create a function using the pycountry_convert package
def country_to_continent(country_name):
    try:
        country_alpha2 = pc.country_name_to_country_alpha2(country_name)
        country_continent_code = pc.country_alpha2_to_continent_code(country_alpha2)
        country_continent_name = pc.convert_continent_code_to_continent_name(country_continent_code)
        return country_continent_name
    except KeyError:
        return 'N/A'


# Create a new data set from happiness that includes a continent column
continent_data = happiness.copy()

continent_data['continent'] = continent_data['Country name'].apply(country_to_continent)
continent_data.head()

,Country name,year,Life Ladder,Log GDP per capita,Social support,Healthy life expectancy at birth,Freedom to make life choices,Generosity,Perceptions of corruption,Positive affect,Negative affect,continent
0,Afghanistan,2008,3.724,7.350,0.451,50.5,0.718,0.168,0.882,0.414,0.258,Asia
1,Afghanistan,2009,4.402,7.509,0.552,50.8,0.679,0.191,0.850,0.481,0.237,Asia
2,Afghanistan,2010,4.758,7.614,0.539,51.1,0.600,0.121,0.707,0.517,0.275,Asia
3,Afghanistan,2011,3.832,7.581,0.521,51.4,0.496,0.164,0.731,0.480,0.267,Asia
4,Afghanistan,2012,3.783,7.661,0.521,51.7,0.531,0.238,0.776,0.614,0.268,Asia


In [73]:
# Find the mean Life Ladder for each continent
continent_grouped = continent_data.groupby(['year', 'continent']).mean(numeric_only = True)

continent_grouped = continent_grouped.reset_index()

# Remove the N/A observations
continent_grouped = continent_grouped[continent_grouped['continent'] != 'N/A']

# Remove years before 2008
continent_grouped = continent_grouped[continent_grouped['year'] >= 2008]

continent_grouped

,year,continent,Life Ladder,Log GDP per capita,Social support,Healthy life expectancy at birth,Freedom to make life choices,Generosity,Perceptions of corruption,Positive affect,Negative affect
21,2008,Africa,4.314615,7.797615,0.666154,51.211538,0.599400,-0.005077,0.843308,0.635846,0.238192
22,2008,Asia,5.112031,9.065375,0.736969,62.701250,0.680387,0.030750,0.794967,0.596531,0.249906
23,2008,Europe,6.608636,10.485091,0.920136,67.580000,0.767091,0.003238,0.669273,0.670810,0.210955
25,2008,North America,6.074167,9.462083,0.857583,61.215000,0.749583,0.099417,0.769333,0.759083,0.228583
26,2008,Oceania,7.317500,10.625500,0.945500,69.900000,0.904500,0.297500,0.382500,0.756500,0.225000
...,...,...,...,...,...,...,...,...,...,...,...
119,2022,Asia,5.252826,9.598143,0.777913,66.127174,0.794955,0.057000,0.756619,0.593130,0.303870
120,2022,Europe,6.467765,10.593061,0.902618,69.932353,0.832206,0.040939,0.631794,0.658235,0.232441
122,2022,North America,6.369000,9.701909,0.847182,66.547727,0.864182,-0.012455,0.726091,0.774909,0.277455
123,2022,Oceania,7.005000,10.768000,0.949000,70.775000,0.842500,0.170000,0.413000,0.708500,0.227000


In [74]:
# Now we can finally plot life ladder per year for every continent!

continents = alt.Chart(continent_grouped, title = 'Continent Life Ladder per Year').mark_line().encode(
    y = alt.Y('Life Ladder', scale = alt.Scale(domain = [4, 7.5])),
    x = alt.X('year', scale = alt.Scale(domain = [2008, 2022])),
    color = 'continent'
)

continents

alt.Chart(...)

In [75]:
# Oceania has consistently been the happiest, Europe, North America, and South America have
# battled it out but today it goes Europe, North America, South America, and Asia and Africa have been
# a consistent 2nd to last and last place in terms of happiness.

In [76]:
# Once again, let's see how this correlates to GDP.
continents_gdp = alt.Chart(continent_grouped, title = 'Continent Log GDP per capita per Year').mark_line().encode(
    y = alt.Y('Log GDP per capita', scale = alt.Scale(domain = [7.5, 11])),
    x = alt.X('year', scale = alt.Scale(domain = [2008, 2022])),
    color = 'continent'
)

continents_gdp

alt.Chart(...)

In [77]:
# Order: Oceania, Europe, North America, South America, Asia, Africa. Exact same order!

In [78]:
##### EXPLORATORY ANALYSIS #####
######## QUESTION FOUR ########
################################

# (iv) Does Life Expectancy influence Life Ladder? 
# Which country has the lowest Life Expectancy and which has the highest? Continents? Compare their Life Ladder values.

# Create a plot of Life Expectancy vs Life Ladder per year
le_vs_ll = alt.Chart(happiness, title = 'Life Ladder vs Life Expectancy').mark_point(opacity = 0.8).encode(
    x = alt.X('Life Ladder', scale = alt.Scale(domain = [1, 8.2])),
    y = alt.Y('Healthy life expectancy at birth', scale = alt.Scale(domain = [0, 80])),
    color = 'year'
)

le_vs_ll

alt.Chart(...)

In [79]:
# There are a few outliers. Let's find what they are.
happiness_ascending_life_exp = happiness.sort_values('Healthy life expectancy at birth', ascending = True)

happiness_ascending_life_exp.head()

,Country name,year,Life Ladder,Log GDP per capita,Social support,Healthy life expectancy at birth,Freedom to make life choices,Generosity,Perceptions of corruption,Positive affect,Negative affect
766,Haiti,2006,3.754,7.973,0.694,6.72,0.449,0.359,0.854,0.583,0.332
767,Haiti,2008,3.846,8.012,0.679,17.36,0.465,0.217,0.812,0.573,0.256
768,Haiti,2010,3.766,7.984,0.554,28.00,0.373,0.171,0.848,0.495,0.293
769,Haiti,2011,4.845,8.022,0.567,33.32,0.413,0.198,0.682,0.550,0.245
770,Haiti,2012,4.413,8.012,0.749,38.64,0.482,0.246,0.717,0.557,0.284


In [80]:
# All of the outliers are Haiti. For reasons we know

In [81]:
# Find the country with the highest average life expectancy and lowest.
# Do the same for continents.

# Finding lowest average life expectancy
grouped_countries = happiness.groupby('Country name').mean(numeric_only = True)
grouped_countries = grouped_countries.reset_index()
grouped_countries_lowest = grouped_countries.sort_values('Healthy life expectancy at birth', ascending = True).iloc[0]
grouped_countries_lowest

Country name                              Haiti
year                                2012.727273
Life Ladder                            3.954182
Log GDP per capita                     8.029818
Social support                         0.616182
Healthy life expectancy at birth      39.770909
Freedom to make life choices           0.461636
Generosity                             0.264091
Perceptions of corruption              0.752091
Positive affect                        0.558182
Negative affect                        0.313182
Name: 59, dtype: object

In [82]:
# Finding highest average life expectancy country
grouped_countries_highest = grouped_countries.sort_values('Healthy life expectancy at birth', ascending = False).iloc[0]
grouped_countries_highest

Country name                              Japan
year                                2013.941176
Life Ladder                            6.030294
Log GDP per capita                    10.589294
Social support                         0.902353
Healthy life expectancy at birth      73.481176
Freedom to make life choices           0.803294
Generosity                            -0.175067
Perceptions of corruption                 0.692
Positive affect                        0.696765
Negative affect                        0.181588
Name: 73, dtype: object

In [83]:
# Just as expected. Japan is highest and Haiti is lowest

In [84]:
# Average life expectancy per continent
continent_life_exp = continent_data.groupby('continent').mean(numeric_only = True)
continent_life_exp = continent_life_exp.reset_index()

# Remove the N/A observations
continent_life_exp = continent_life_exp[continent_life_exp['continent'] != 'N/A']

# Make a bar chart
bar_cont_life = alt.Chart(continent_life_exp, title = 'Average Life Expectancy per Continent').mark_bar(opacity = 0.8).encode(
    x = 'continent',
    y = 'Healthy life expectancy at birth',
    color = 'continent'
).properties(
    width = 500,
    height = 300
)

bar_cont_life

alt.Chart(...)

In [85]:
# Order: Oceania > Europe > South America > North America > Asia > Africa

In [86]:
##### EXPLORATORY ANALYSIS #####
######## QUESTION FIVE ########
################################

# (v) How do social support, freedom to make life choices, generosity, perceptions of corruption, positive affect,
# and negative affect correlate with Life Ladder?
# Do any of them correlate with each other? Why do you think this is?

# Create correlation matrix
happiness_corr = happiness.drop(columns = ['Country name', 'year'])
corr_mx = happiness_corr.corr()

# Melt corr_mx
corr_mx_long = corr_mx.reset_index().rename(
    columns = {'index': 'row'}
).melt(
    id_vars = 'row',
    var_name = 'col',
    value_name = 'Correlation'
)

corr_mx_long

,row,col,Correlation
0,Life Ladder,Life Ladder,1.000000
1,Log GDP per capita,Life Ladder,0.784868
2,Social support,Life Ladder,0.721662
3,Healthy life expectancy at birth,Life Ladder,0.713499
4,Freedom to make life choices,Life Ladder,0.534493
...,...,...,...
76,Freedom to make life choices,Negative affect,-0.275438
77,Generosity,Negative affect,-0.080801
78,Perceptions of corruption,Negative affect,0.266265
79,Positive affect,Negative affect,-0.330301


In [87]:
# Construct a heatmap
alt.Chart(corr_mx_long, title = 'Happiness Correlations').mark_rect().encode(
    x = alt.X('col', title = '', sort = alt.SortField('Correlation', order = 'ascending')),
    y = alt.Y('row', title = '', sort = alt.SortField('Correlation', order = 'ascending')),
    color = alt.Color('Correlation',
                    scale = alt.Scale(scheme = 'blueorange',
                                      domain = (-1, 1),
                                      type = 'sqrt'),
                    legend = alt.Legend(tickCount = 5))
).properties(width = 300, height = 300)

alt.Chart(...)

In [88]:
# The correlations varied which we will discuss in our summary. 

## (ii) Regression Analysis

Using life ladder as our dependent variable, we can explore which variables affect the outcome the most. Linear regression is an appropriate model for this type of data considering the fact that our one dependent variable likely has a very direct relationship with the independent variables. Using multiple linear regression would be appropriate 
for a model which includes more than one independent variable. We can also fit other models which are univariate relationships between one variable and our dependent variable and explore these relationships. 

In [89]:
#in order to have clean data for our regression model, we will drop na values

clean_happiness_data = happiness.dropna()

# our dependent variable
y = clean_happiness_data['Life Ladder']

# constructing the explanatory variable matrix using social support as our independent variable
x = sm.tools.add_constant(clean_happiness_data.loc[:, 'Social support'])



#fitting simple regression model
simple_regression = sm.OLS(endog = y, exog = x)

# store results
rslt = simple_regression.fit() 



In [90]:
# print several rows of explanatory variable matrix
x.head()

,const,Social support
0,1.0,0.451
1,1.0,0.552
2,1.0,0.539
3,1.0,0.521
4,1.0,0.521


In [91]:
rslt.params

const             0.059105
Social support    6.681637
dtype: float64

In [92]:
rslt.cov_params()

,const,Social support
const,0.013916,-0.016827
Social support,-0.016827,0.020823


In [93]:
rslt.rsquared

0.522925254307458

When looking at the R squared value for our model, we see that it falls around 0.52 which means that this model is not particularly strong at explaining variation of the response variable. 

In [94]:
clean_happiness_data.loc[:,'fitted_value'] = rslt.fittedvalues
clean_happiness_data.loc[:,'residual'] = rslt.resid

/tmp/ipykernel_81/1753417206.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  clean_happiness_data.loc[:,'fitted_value'] = rslt.fittedvalues
/tmp/ipykernel_81/1753417206.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  clean_happiness_data.loc[:,'residual'] = rslt.resid


In [95]:
#scatter plot to visually represent this relationhsip 

scatter_happy = alt.Chart(clean_happiness_data).mark_circle().encode(
    y = alt.Y('Life Ladder', scale = alt.Scale(zero = False), title = 'Life Ladder Score'),
    x = alt.X('Social support', scale = alt.Scale(zero = False), 
              title = 'Social support score')
    
)
scatter_happy

alt.Chart(...)

In [96]:
#Now we are going to fit another simple regression model which looks at Life Ladder versus Log Gdp

x1 = sm.tools.add_constant(clean_happiness_data.loc[:, 'Log GDP per capita'])

simple_regression_2 = sm.OLS(endog = y, exog = x1)

rslt2 = simple_regression_2.fit()

In [97]:
x1.head()

,const,Log GDP per capita
0,1.0,7.350
1,1.0,7.509
2,1.0,7.614
3,1.0,7.581
4,1.0,7.661


In [98]:
rslt2.params

const                -1.796023
Log GDP per capita    0.776397
dtype: float64

In [99]:
rslt2.cov_params()

,const,Log GDP per capita
const,0.016683,-0.001758
Log GDP per capita,-0.001758,0.000188


In [100]:
coef_tbl = pd.DataFrame({'estimate': rslt2.params.values,
              'standard error': np.sqrt(rslt2.cov_params().values.diagonal())},
              index = x1.columns)
coef_tbl.loc['error variance', 'estimate'] = rslt2.scale

coef_tbl

,estimate,standard error
const,-1.796023,0.129162
Log GDP per capita,0.776397,0.013718
error variance,0.493851,NaN


In [101]:
rslt2.rsquared

0.6208702719354315

In [102]:
clean_happiness_data.loc[:,'fitted_value_2'] = rslt2.fittedvalues
clean_happiness_data.loc[:,'residual_2'] = rslt2.resid
clean_happiness_data

# con

/tmp/ipykernel_81/371521334.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  clean_happiness_data.loc[:,'fitted_value_2'] = rslt2.fittedvalues
/tmp/ipykernel_81/371521334.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  clean_happiness_data.loc[:,'residual_2'] = rslt2.resid


,Country name,year,Life Ladder,Log GDP per capita,Social support,Healthy life expectancy at birth,Freedom to make life choices,Generosity,Perceptions of corruption,Positive affect,Negative affect,fitted_value,residual,fitted_value_2,residual_2
0,Afghanistan,2008,3.724,7.350,0.451,50.500,0.718,0.168,0.882,0.414,0.258,3.072523,0.651477,3.910498,-0.186498
1,Afghanistan,2009,4.402,7.509,0.552,50.800,0.679,0.191,0.850,0.481,0.237,3.747368,0.654632,4.033945,0.368055
2,Afghanistan,2010,4.758,7.614,0.539,51.100,0.600,0.121,0.707,0.517,0.275,3.660507,1.097493,4.115467,0.642533
3,Afghanistan,2011,3.832,7.581,0.521,51.400,0.496,0.164,0.731,0.480,0.267,3.540237,0.291763,4.089846,-0.257846
4,Afghanistan,2012,3.783,7.661,0.521,51.700,0.531,0.238,0.776,0.614,0.268,3.540237,0.242763,4.151958,-0.368958
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
2194,Zimbabwe,2018,3.616,7.783,0.775,52.625,0.763,-0.051,0.844,0.658,0.212,5.237373,-1.621373,4.246678,-0.630678
2195,Zimbabwe,2019,2.694,7.698,0.759,53.100,0.632,-0.047,0.831,0.658,0.235,5.130467,-2.436467,4.180684,-1.486684
2196,Zimbabwe,2020,3.160,7.596,0.717,53.575,0.643,0.006,0.789,0.661,0.346,4.849838,-1.689838,4.101492,-0.941492
2197,Zimbabwe,2021,3.155,7.657,0.685,54.050,0.668,-0.076,0.757,0.610,0.242,4.636026,-1.481026,4.148852,-0.993852


In [103]:
#create a line plot using our new fitted values
slr_line = alt.Chart(clean_happiness_data).mark_line().encode(
    x = 'Log GDP per capita',
    y = 'fitted_value_2'
)

# layer using our scatter plot from earlier in our exploratory analysis

ll_vs_gdp + slr_line

alt.LayerChart(...)

### Multiple Regression

In [104]:
clean_happiness_data.columns

Index(['Country name', 'year', 'Life Ladder', 'Log GDP per capita',
       'Social support', 'Healthy life expectancy at birth',
       'Freedom to make life choices', 'Generosity',
       'Perceptions of corruption', 'Positive affect', 'Negative affect',
       'fitted_value', 'residual', 'fitted_value_2', 'residual_2'],
      dtype='object')

In [105]:
#Here we want to change our model to fit more than one independent variable
#Our variables are Healthy life expectancy at birth, Generosity, and Freedom to make life choices
x = sm.tools.add_constant(clean_happiness_data.loc[:, ['Healthy life expectancy at birth', 'Freedom to make life choices', 'Log GDP per capita']])

# fiting model and storing results
mlr = sm.OLS(endog = y, exog = x)
rslt = mlr.fit() 

#constructing a coefficient table to see how variables are related
coef_tbl_mlr = pd.DataFrame({
    'estimate': rslt.params.values,
    'standard error': np.sqrt(rslt.cov_params().values.diagonal()) 
    },
    index = x.columns 
)

coef_tbl_mlr.loc['error variance', 'estimate'] = rslt.scale



In [106]:
coef_tbl_mlr

,estimate,standard error
const,-2.969754,0.131358
Healthy life expectancy at birth,0.026103,0.003627
Freedom to make life choices,2.220309,0.108536
Log GDP per capita,0.548450,0.022037
error variance,0.390357,NaN


In [107]:
rslt.rsquared

0.7006290577321692

The above coefficients, error values, and R squared value are interesting. When we include Healthy life expecatncy from birth as well as Freedom to make life choices on top of the Log GDP per capita we see that Log GDP per capita has a slightly lower coefficent value and higher standard error. Additionally, the other variables in the multiple regression led to a higher R squared value of 0.7 which indicates that more of the variance of our dependent variable is being explained by including the other variables compared to log GDP alone. 

## Summary of Findings

Upon conducting extensive exploratory data analysis as well as simple and multiple linear regression we were able to answer the questions that we originally posed when looking at the data. 
1. In looking at the relationship between life ladder and log GDP per capita, we found a fairly strong positive relationship. This was unsurprising but there were several interesting findings through both exploratory techniques as well as regression analysis. Through a basic scatterplot we saw a strong positive correlation which had higher gdp values in recent years. Our simple regression model had a positive coefficient of 0.77 for log gdp compared to life ladder score which corroborated our findings  
2. We posed the question of whether certain countries had experienced a large change in life ladders in the years observed. The countries with a positive change were: Congo, El Salvador, Georgia, Lithuania, Mongolia, Nepal, Nicaragua, Taiwan, Togo, Uruguay. The countries with a negative change were: Afghanistan, Bangladesh, Botswana, India, Lebanon. There tended to be a strong correlation with negative change and decreasing gdp and vice versa. With this said, there are likely other political/social events that could have caused these change such as war or political turmoil. 
3. We wanted to explore whether happiness varied by continent. Through groupings and mean estimates we found Oceania has consistently been the happiest. Europe, South America and North America have alternated for the second happiest continent. By the most recent measurement 2nd place is Europe, 3rd is North America, 4th is South America. Throughout the entire time frame, Asia has been 5th and Africa 6th. This was fascinating and we can infer that political stability, and general peace of the continents likely factor into these mean estimates.
4. The next relationship we explored was between life expectancy and life ladder score. We looked at the countries with highest and lowest life expectancy over the time frame which were Japan and Haiti respectively. The 2010 Haiti earthquake certainly had an effect on this estimate. Japan certainly had a higher life ladder score than Haiti. Interestingly, the life expectancy by continent followed the exact same order as the happiness by continent from highest to lowest.
5. Our final question was: How do social support, freedom to make life choices, generosity, perceptions of corruption, positive affect, and negative affect correlate with Life Ladder? By making a correlation matrix and constructing a heatmap we found that Log gdp per capita, social support, healthy life expectancy and freedom to make life choices all had positive correlations above 0.5. Negative affect and Perceptions of corruption both had a negative correlation with life ladder score which makes sense. 

Ultimately, exploring geographical trends of happiness and looking into which factors affect the overall level of happiness for people in different countries was very interesting. It provided us with opportunities think about how real-world events can have an effect on people's well-being and their happiness. 